lib/modem_jwt/modem_jwt.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+#include <zephyr.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <modem/at_cmd.h>
+#include <modem/modem_jwt.h>
+
+#define JWT_CMD_TEMPLATE "AT%%JWT=%d,%u,\"%s\",\"%s\",%d,%d\r\n"
+#define MAX_INT_PRINT_SZ 11
+#define JWT_CMD_PRINT_INT_SZ (MAX_INT_PRINT_SZ * 4)
+
+/* Minimum JWT response is ~420 bytes.
+ * With very long (128 bytes) values for sub and aud claims
+ * response is ~820.
+ */
+BUILD_ASSERT(CONFIG_MODEM_JWT_MAX_LEN <= CONFIG_AT_CMD_RESPONSE_MAX_LEN,
+	     "Max JWT length exceeds max AT cmd response length");
+
+static int parse_jwt_at_cmd_resp(char *const jwt_resp)
+{
+	if (!jwt_resp) {
+		return -EINVAL;
+	}
+
+	char *jwt_start = NULL;
+	char *jwt_end = NULL;
+	size_t jwt_len;
+
+	/* Response format is:
+	 *	%JWT: "<base64_JWT>"\r\n
+	 */
+
+	/* A quotation mark indicates the start of the JWT string */
+	jwt_start = strchr(jwt_resp, '"');
+	if (!jwt_start) {
+		return -EBADMSG;
+	}
+
+	/* Move beyond first quotation mark and find second */
+	jwt_end = strchr(++jwt_start, '"');
+	if (!jwt_end) {
+		return -EBADMSG;
+	}
+
+	*jwt_end = 0;
+	jwt_len = strlen(jwt_start) + 1;
+
+	/* Move the now null-terminated jwt to the front of the buffer */
+	memmove(jwt_resp, jwt_start, jwt_len);
+
+	return 0;
+}
+
+int modem_jwt_generate(struct jwt_data *const jwt)
+{
+	if (!jwt) {
+		return -EINVAL;
+	}
+
+	int ret = 0;
+	enum at_cmd_state state;
+	char *cmd;
+	char *cmd_resp = NULL;
+	size_t cmd_sz = sizeof(JWT_CMD_TEMPLATE) + JWT_CMD_PRINT_INT_SZ +
+			(jwt->subject ? strlen(jwt->subject) : 0) +
+			(jwt->audience ? strlen(jwt->audience) : 0);
+
+	jwt->jwt_out = NULL;
```

Comment:
Could it be an option to pass a buffer as an input argument in `jwt_out` (together with a buffer size)? If `jwt_out` is `NULL`, the current behavior of allocating on the heap is kept. If a buffer is passed in `jwt_out`, the JWT is stored there instead. Just wondering if there are use cases where it's more convenient for users to allocate memory statically thmselves instead of using the heap, but I might be over-complicating it

Reply:
sounds like a good option